spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/MigrateDeltaLakeTableSparkAction.java

```diff
@@ -0,0 +1,311 @@
+              partitionSpec,
+              partitions
+      );
+
+      stagedTable.commitStagedChanges();
+      error = false;
+    } finally {
+      if (error) {
+        LOG.error("Failed to perform the migration, aborting table creation and restoring the original table");
+        restoreSourceTable(destCatalog, backupIdent, identifier);
+        if (stagedTable != null) {
+          try {
+            stagedTable.abortStagedChanges();
+          } catch (Exception abortException) {
+            LOG.error("Cannot abort staged changes", abortException);
+          }
+        }
+      }
+    }
+    Snapshot snapshot = icebergTable.currentSnapshot();
+    long numFilesMigrated = Long.parseLong(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+    LOG.info("Successfully loaded Iceberg metadata for {} files to {}", numFilesMigrated, identifier);
+    return new BaseMigrateDeltaLakeTableActionResult(numFilesMigrated);
+  }
+
+  private static void restoreSourceTable(StagingTableCatalog destinationCatalog,
+                                  Identifier backupIdent, Identifier sourceTableIdent) {
+    try {
+      destinationCatalog.renameTable(backupIdent, sourceTableIdent);
+    } catch (org.apache.spark.sql.catalyst.analysis.NoSuchTableException e) {
+      LOG.error("Cannot restore the original table, the backup table {} cannot be found", backupIdent, e);
+
+    } catch (org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException e) {
+      LOG.error("Cannot restore the original table, a table with the original name exists. " +
+              "Use the backup table {} to restore the original table manually.", backupIdent, e);
+    }
+  }
+
+  private static List<SparkTableUtil.SparkPartition> getSparkPartitionsFromDeltaSnapshot(
+      io.delta.standalone.Snapshot updatedSnapshot,
+      Path deltaLogPath
+  ) {
+    return updatedSnapshot.getAllFiles()
+        .stream()
+        // Map each partition to the list of files within it
+        .collect(Collectors.groupingBy(AddFile::getPartitionValues))
+        .entrySet()
+        .stream()
+        .map(entry -> {
+              // We don't care what value we take since they will all have the same prefix.
+              // The arbitrary file will have a path that looks like "partition1/partition2/file.parquet,
+              // We're interested in the part prior to the filename
+              AddFile addFile = entry.getValue().get(0);
+              String pathBeforeFileName = addFile.getPath().substring(0, addFile.getPath().lastIndexOf("/"));
+              String fullPath = new Path(deltaLogPath, pathBeforeFileName).toString();
+
+              return new SparkTableUtil.SparkPartition(
+                  entry.getKey(), // Map containing name and values of partitions
+                  fullPath,
+                  // Delta tables only support parquet
+                  "parquet"
+              );
+        }
+        )
+        .collect(Collectors.toList());
+  }
```

Comment:
Should we parallelize this computation? For example, concurrently across files update a grouping based on partition values. Could be overkill (depends on the number of files in the snapshot)